File: HM/1312_MinimumInsertionStepsToMakeAStringPalindrome.py
#Longest Palindromic Subsequence를 찾아서 그거를 전체 길이에서 빼주면 끝! 매우쉽다.
def minInsertions(s):
        """
        :type s: str
        :rtype: int
        """
            
        def longestPalindromeSubseq(s):
            s2 = s[::-1]
            #[row[col]] = [s[s2]]
            dp = [[0 for j in range(len(s2)+1)]for i in range(len(s)+1)] 

            for i in range(len(s) -1, -1, -1):
                for j in range(len(s2) -1, -1, -1):
                    if s[i] == s2[j]:
                        dp[i][j] = 1 + dp[i+1][j+1] #diagnoal에 1을 더해주기
                    else:
                        dp[i][j] = max(dp[i][j+1], dp[i+1][j]) #matching 하지 않으면 오른쪽이나 밑에 값 중 더 큰 것을 가져오기

            return dp[0][0]
        
        lengthLongestPalindrome = longestPalindromeSubseq(s)
        # 최소 삽입 횟수는 전체 길이에서 최장 회문 부분 수열의 길이를 뺀 값
        min_insertions = len(s) - lengthLongestPalindrome
        return min_insertions

File: HM/1312_MinimumInsertionStepsToMakeAStringPalindrome.c
#python과 동일하게 longest palindromic subsequence의 길이를 찾아서 string length에서 빼준다.

int minInsertions(char* s) {
    int lengthLongestPalindrome = longestPalindromeSubseq(s);
    // 최소 삽입 횟수는 전체 길이에서 최장 회문 부분 수열의 길이를 뺀 값
    int min_insertions = strlen(s) - lengthLongestPalindrome;
    return min_insertions;
}

int max(int a, int b) {
    return (a > b) ? a : b;
}

int longestPalindromeSubseq(char* s) {
    int n = strlen(s);
    char s2[n + 1];
    
    // Reverse the string s
    for (int i = 0; i < n; ++i) {
        s2[i] = s[n - 1 - i];
    }
    s2[n] = '\0';  // Null-terminate the reversed string
    
    // Create a 2D array for dynamic programming
    int dp[n + 1][n + 1];

    for (int i = 0; i <= n; ++i) {
        for (int j = 0; j <= n; ++j) {
            dp[i][j] = 0;
        }
    }

    // Fill the dp array using dynamic programming
    for (int i = n - 1; i >= 0; --i) {
        for (int j = n - 1; j >= 0; --j) {
            if (s[i] == s2[j]) {
                dp[i][j] = 1 + dp[i + 1][j + 1];
            } else {
                dp[i][j] = max(dp[i][j + 1], dp[i + 1][j]);
            }
        }
    }

    return dp[0][0];
}
